src/sentry/tasks/releasemonitor.py

```diff
@@ -0,0 +1,165 @@
+import logging
+import time
+from datetime import timedelta
+
+from django.db.models import Subquery
+from django.utils import timezone
+
+from sentry.models import Environment, Release, ReleaseProjectEnvironment
+from sentry.tasks.base import instrumented_task
+from sentry.utils import metrics
+from sentry.utils.snuba import Dataset, raw_query
+
+REQUIRED_ADOPTION_PERCENT = 0.1
+CHUNK_SIZE = 1000
+MAX_SECONDS = 60
+
+logger = logging.getLogger("tasks.releasemonitor")
+
+
+@instrumented_task(
+    name="sentry.tasks.monitor_release_adoption",
+    queue="releasemonitor.monitor_release_adoption",
+    default_retry_delay=5,
+    max_retries=5,
+)
+def monitor_release_adoption(**kwargs):
+    metrics.incr("sentry.tasks.monitor_release_adoption.start")
+    # 1. Runs
+    # SELECT org_id, uniq(project_id)
+    # FROM sessions_hourly_dist
+    # WHERE date > now() - interval '<whatever>'
+    # GROUP BY org_id
+    # Date range here will be 12 hours to start.
+    # NOTE: When this query is supported, make sure you paginate it.
+    # NOTE: Hardcoded data for sentry org and sentry project for early release, in the same format snuba should return
+    data = [
+        {"org_id": [1], "project_id": [1]},
+    ]
+
+    # NOTE: This should probably be broken out into a separate task per org because it potentially has to paginate through a lot of snuba results.
+    with metrics.timer("sentry.tasks.monitor_release_adoption.process_projects_with_sessions"):
+        for row in data:
+            process_projects_with_sessions(data["org_id"][0], data["project_id"])
+
+
+def process_projects_with_sessions(org_id, project_ids):
+    # Accepts a single org id and a list of project ids
+    # 2. For each org result from #1, run a query similar to
+    # SELECT
+    #     release AS _snuba_release,
+    #     environment AS _snuba_environment,
+    #     project AS _snuba_project,
+    #     countIfMerge(sessions) + sumIfMerge(sessions_preaggr) AS _snuba_sessions
+    # FROM sessions_hourly_local
+    # WHERE org_id = <org id>
+    # AND project_id IN (<the projects for that org>)
+    # AND started >= toDateTime(<however long ago>, 'Universal')
+    # AND started < toDateTime(<now>, 'Universal')
+    # GROUP BY
+    #     _snuba_release,
+    #     _snuba_environment,
+    #     _snuba_project,
+    # LIMIT 0, 1000
+    # Date range here is 6 hours.
+    time_start = time.time()
+    offset = 0
+    total_sessions = 0
+    updated_ids = []
+    with metrics.timer("sentry.tasks.monitor_release_adoption.process_projects_with_sessions.loop"):
+        while (time.time() - time_start) < MAX_SECONDS:
```

Comment:
If we exceed this we should probably log something, either a metric or an error in sentry (or both). Also wondering whether we still need this limit if this call becomes its own task?